# A1A2resequencing
Scripts used in analyzing diversity for _G. herbaceum_ and _G. arboreum_, the A-genome cottons

_all.exon_dxy.txt.gz_: estimate of between population nucleotide divergence in exons

_all.exon_fst.txt.gz_: estimate of Weir and Cockerham’s FST in exons

_all.exon_pi.txt.gz_: estimate of nucleotide diversity in exons

_all.gene_dxy.txt.gz_: estimate of between population nucleotide divergence in exons

_all.gene_fst.txt.gz_: estimate of Weir and Cockerham’s FST in exons

_all.gene_pi.txt.gz_: estimate of nucleotide diversity in exons

_all.intergenic_dxy.txt.gz_: estimate of between population nucleotide divergence in exons

_all.intergenic_fst.txt.gz_: estimate of Weir and Cockerham’s FST in exons

_all.intergenic_pi.txt.gz_: estimate of nucleotide diversity in exons

_all.intron_dxy.txt.gz_: estimate of between population nucleotide divergence in exons

_all.intron_fst.txt.gz_: estimate of Weir and Cockerham’s FST in exons

_all.intron_pi.txt.gz_: estimate of nucleotide diversity in exons

_pixy_dxy.txt.gz_: estimate of between population nucleotide divergence in exons

_pixy_fst.txt.gz_: estimate of Weir and Cockerham’s FST in exons

_pixy_pi.txt.gz_: estimate of nucleotide diversity in exons

_longicalyx.exon.bed_: bed file of exon positions

_longicalyx.gene.bed_: bed file of gene positions

_longicalyx.intergenic.bed_: bed file of intergenic positions

_longicalyx.intron.bed_: bed file of intron positions

_pi.R_: R script to summarize pixy statistics

_PixyPiDiversity.jpg_: image similar to Figure 1A, but made with pixy diversity estimates

_populations.bcftools_: population file for pixy