# A1A2resequencing
Scripts used in analyzing diversity using vcftools in _G. herbaceum_ and _G. arboreum_, the A-genome cottons


_A1.exon.piDiversity.windowed.pi_: nucleotide diversity in exons for _G._ _herbaceum_

_A1.intergenic.piDiversity.windowed.pi_: nucleotide diversity in intergenic space for _G._ _herbaceum_

_A1.intron.piDiversity.windowed.pi_: nucleotide diversity in introns for _G._ _herbaceum_

_A1.piDiversity.windowed.pi_: nucleotide diversity in _G._ _herbaceum_

_A1vA2.fst.larger.windowed.weir.fst_: Windowed Weir and Cockerham Fst in 10 Mb windows

_A1vA2.fst.large.windowed.weir.fst_: Windowed Weir and Cockerham Fst in 1 Mb windows

_A1vA2.fst.largish.windowed.weir.fst_: Windowed Weir and Cockerham Fst in 5 Mb windows

_A1vA2.fst.windowed.weir.fst_: Windowed Weir and Cockerham Fst in 0.1 Mb windows

_A2.exon.piDiversity.windowed.pi_: nucleotide diversity in exons for _G._ _arboreum_

_A2.intergenic.piDiversity.windowed.pi_: nucleotide diversity in intergenic space for _G._ _arboreum_

_A2.intron.piDiversity.windowed.pi_: nucleotide diversity in introns for _G._ _arboreum_

_A2.piDiversity.windowed.pi_: nucleotide diversity in _G._ _arboreum_

_diversity.distribution.R_: R script to summarize diversity

_diversity.vcftools.slurm_: slurm script to run vcftools

_Fst.R_: R script to summarize Fst

_longicalyx.exon.bed_: bed file of exon positions

_longicalyx.gene.bed_: bed file of gene positions

_longicalyx.intergenic.bed_: bed file of intergenic positions

_longicalyx.intron.bed_: bed file of intron positions

